pulsar-io/kafka/src/main/java/org/apache/pulsar/io/kafka/KafkaBytesSource.java

```diff
@@ -43,14 +49,64 @@
 
     @Override
     protected Properties beforeCreateConsumer(Properties props) {
-        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
-        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+        props.putIfAbsent("schema.registry.url", "http://localhost:8081");
+        props.putIfAbsent(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+        props.putIfAbsent(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+
+        String currentValue = props.getProperty(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG);
+
+        // replace KafkaAvroDeserializer with our custom implementation
+        if (currentValue != null && currentValue.equals(KafkaAvroDeserializer.class.getName())) {
+            props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, NoCopyKafkaAvroDeserializer.class.getName());
+        }
+
         log.info("Created kafka consumer config : {}", props);
         return props;
     }
 
     @Override
-    public byte[] extractValue(ConsumerRecord<String, byte[]> record) {
-        return record.value();
+    public Object extractValue(ConsumerRecord<String, Object> consumerRecord) {
+        Object value = consumerRecord.value();
+        if (value instanceof BytesWithAvroPulsarSchema) {
+            return ((BytesWithAvroPulsarSchema) value).getValue();
+        }
+        return value;
+    }
+
+    @Override
+    public org.apache.pulsar.client.api.Schema<?> extractSchema(ConsumerRecord<String, Object> consumerRecord) {
+        Object value = consumerRecord.value();
+        if (value instanceof BytesWithAvroPulsarSchema) {
+            return ((BytesWithAvroPulsarSchema) value).getPulsarSchema();
+        } else {
+            return org.apache.pulsar.client.api.Schema.BYTES;
+        }
+    }
+
+    public static class NoCopyKafkaAvroDeserializer extends KafkaAvroDeserializer {
+
+        private final PulsarSchemaCache<GenericRecord> schemaCache = new PulsarSchemaCache<>();
+
+        @Override
+        protected Object deserialize(boolean includeSchemaAndVersion, String topic, Boolean isKey, byte[] payload, Schema readerSchema) throws SerializationException {
+            if (payload == null) {
+                return null;
+            } else {
+                int id = -1;
+                try {
+                    ByteBuffer buffer = ByteBuffer.wrap(payload);
+                    buffer.get(); // magic number
+                    id = buffer.getInt();
+                    String subject = getSubjectName(topic, isKey != null ? isKey : false);
+                    Schema schema = this.schemaRegistry.getBySubjectAndId(subject, id);
```

Comment:
I think getting request schema information should be done in the `extractSchema` call, not in a deserializer.

Reply:
The `schemaRegistry` has its own cache (it should be a  `CachedSchemaRegistryClient`)

This version of the code is a simplified version of the original `deserialize` method.
I wanted to keep it as close as possible to the original implementation.

```
protected Object deserialize(boolean includeSchemaAndVersion, String topic, Boolean isKey, byte[] payload, Schema readerSchema) throws SerializationException {
        if (payload == null) {
            return null;
        } else {
            byte id = -1;

            try {
                ByteBuffer buffer = this.getByteBuffer(payload);
                int id = buffer.getInt();
                String subject = includeSchemaAndVersion ? getSubjectName(topic, isKey) : null;
                Schema schema = this.schemaRegistry.getBySubjectAndId(subject, id);
                int length = buffer.limit() - 1 - 4;
                Object result;
```

Reply:
In the end we have to pass a byte[] to the PulsarProducer as value for the Record (we are implementing a `Source<byte[]>`, so we have to copy this data at least once.

There is no way to save this memory copy, because we have to skip the magic number and the schema id.

Reply:
Why not implement `Source<ByteBuf>` or `Source<ByteBuffer>`?